File: internal/domain/series.go
package domain

import (
	"fmt"
	"math"
)

// Unit tags a number so the frontend can format it without knowing where it
// came from. Conversion happens here, never in a Svelte template: the reference
// implementation multiplied ALB seconds by 1000 inside the view while app
// latency arrived already in milliseconds, which left two numbers on one screen
// whose units could only be reconciled by reading the markup.
type Unit string

const (
	UnitNone    Unit = ""
	UnitMillis  Unit = "ms"
	UnitSeconds Unit = "s"
	UnitPercent Unit = "%"
	UnitCount   Unit = "count"
	UnitPerSec  Unit = "/s"
	UnitBytes   Unit = "bytes"
	UnitConns   Unit = "conn"
)

// Intent colours a value by what it means, not by which panel it sits on.
type Intent string

const (
	IntentNeutral Intent = "neutral"
	IntentGood    Intent = "good"
	IntentWarn    Intent = "warn"
	IntentBad     Intent = "bad"
)

// Point is one sample. A nil Point is a genuine gap in the data and stays nil
// all the way to uPlot, which renders it as a break in the line. Filling gaps
// with zero would turn "CloudWatch had nothing here" into "the value was zero",
// which reads as a traffic collapse on a latency chart.
type Point = *float64

// P boxes a float into a Point.
func P(v float64) Point { return &v }

// Series is one line on a chart, aligned index-for-index with the enclosing
// payload's Timestamps.
type Series struct {
	Label  string  `json:"label"`
	Unit   Unit    `json:"unit"`
	Color  string  `json:"color,omitempty"`
	Values []Point `json:"values"`
}

// NewSeries allocates a series of n gaps, ready to be filled by index.
func NewSeries(label string, unit Unit, color string, n int) *Series {
	return &Series{Label: label, Unit: unit, Color: color, Values: make([]Point, n)}
}

// Set writes v at i, ignoring out-of-range indices so a stray CloudWatch
// timestamp cannot panic a handler.
func (s *Series) Set(i int, v float64) {
	if i < 0 || i >= len(s.Values) {
		return
	}
	s.Values[i] = P(v)
}

// Add accumulates into bucket i, treating a gap as zero. Use it for counts,
// where several source rows land in the same bucket.
func (s *Series) Add(i int, v float64) {
	if i < 0 || i >= len(s.Values) {
		return
	}
	if s.Values[i] == nil {
		s.Values[i] = P(v)
		return
	}
	*s.Values[i] += v
}

// Defined counts the non-gap samples.
func (s *Series) Defined() int {
	n := 0
	for _, v := range s.Values {
		if v != nil {
			n++
		}
	}
	return n
}

// Max, Min, Sum, Avg and Last reduce a series to a single number, returning nil
// when the series is entirely gaps.
//
// Every headline number on the dashboard goes through one of these. That is the
// point: the overview tile and the detail chart read the same reduction of the
// same series, so they cannot disagree the way the reference implementation's
// hand-rolled client-side reductions did.

func (s *Series) Max() Point { return s.reduce(func(a, b float64) float64 { return math.Max(a, b) }) }
func (s *Series) Min() Point { return s.reduce(func(a, b float64) float64 { return math.Min(a, b) }) }
func (s *Series) Sum() Point { return s.reduce(func(a, b float64) float64 { return a + b }) }

func (s *Series) Avg() Point {
	sum, n := 0.0, 0
	for _, v := range s.Values {
		if v != nil {
			sum += *v
			n++
		}
	}
	if n == 0 {
		return nil
	}
	return P(sum / float64(n))
}

// Last is the newest defined sample. Because NewWindow floors End to a period
// boundary, the final bucket is a complete one — there is no need for the
// "second-to-last bucket" fudge the reference implementation used to dodge its
// own partially-filled tail.
func (s *Series) Last() Point {
	for i := len(s.Values) - 1; i >= 0; i-- {
		if s.Values[i] != nil {
			return s.Values[i]
		}
	}
	return nil
}

func (s *Series) reduce(f func(a, b float64) float64) Point {
	var acc *float64
	for _, v := range s.Values {
		if v == nil {
			continue
		}
		if acc == nil {
			acc = P(*v)
			continue
		}
		*acc = f(*acc, *v)
	}
	return acc
}

// Stat is a single number rendered as text somewhere on the dashboard.
//
// Basis names the population the number was computed over. Two "request count"
// stats drawn from different filters are a real thing that happens — the
// reference implementation had exactly that, one counting rows with a status
// and another counting rows with a latency — and carrying the basis on the wire
// makes the difference visible in the UI instead of leaving it to be discovered
// as a bug.
type Stat struct {
	Key    string `json:"key"`
	Label  string `json:"label"`
	Value  Point  `json:"value"`
	Unit   Unit   `json:"unit"`
	Text   string `json:"text,omitempty"`
	Basis  string `json:"basis,omitempty"`
	Intent Intent `json:"intent,omitempty"`
}

// Column describes one table column, including whether its cells are worth
// offering a copy button for.
type Column struct {
	Key      string `json:"key"`
	Label    string `json:"label"`
	Unit     Unit   `json:"unit,omitempty"`
	Copyable bool   `json:"copyable,omitempty"`
	Mono     bool   `json:"mono,omitempty"`
	Numeric  bool   `json:"numeric,omitempty"`
}

// Row is one table row keyed by Column.Key.
type Row map[string]any

// Table is a list view plus the honest size of the thing it lists.
//
// Total is not derived from Rows. The reference implementation displayed
// `errorLogs.length` as a total while that array was capped at 300 by the SQL
// underneath, so the headline silently stopped counting past 300 and disagreed
// with the aggregate shown beside it. Requiring the caller to supply Total
// separately is what prevents that shape of bug from being written again.
type Table struct {
	Columns   []Column `json:"columns"`
	Rows      []Row    `json:"rows"`
	Total     int64    `json:"total"`
	Truncated bool     `json:"truncated"`
	Limit     int      `json:"limit"`
}

// NewTable pairs rows with a total that was counted independently of them.
func NewTable(cols []Column, rows []Row, total int64, limit int) *Table {
	if rows == nil {
		rows = []Row{}
	}
	return &Table{
		Columns:   cols,
		Rows:      rows,
		Total:     total,
		Truncated: total > int64(len(rows)),
		Limit:     limit,
	}
}

// Bars tells the frontend that a table is really a categorical breakdown and
// which of its columns carry the category and the count.
//
// The backend names the columns rather than letting the view guess, so a bar
// chart and the table beside it are two renderings of one set of rows — not two
// independently derived views that can disagree about what they are counting.
type Bars struct {
	KeyColumn   string `json:"keyColumn"`
	ValueColumn string `json:"valueColumn"`
	// GroupColumn, when set, splits the rows into one breakdown per value.
	GroupColumn string `json:"groupColumn,omitempty"`
}

// Panel is one card on the dashboard: its chart, its headline numbers, and
// optionally a table.
type Panel struct {
	ID       string    `json:"id"`
	Title    string    `json:"title"`
	Series   []*Series `json:"series,omitempty"`
	Stats    []Stat    `json:"stats,omitempty"`
	Table    *Table    `json:"table,omitempty"`
	Bars     *Bars     `json:"bars,omitempty"`
	Warnings []string  `json:"warnings,omitempty"`
}

// Warn appends a caveat the UI should surface rather than swallow.
func (p *Panel) Warn(format string, args ...any) {
	p.Warnings = append(p.Warnings, fmt.Sprintf(format, args...))
}

// WindowJSON is the wire form of a Window. Timestamps live here, once, rather
// than on each series: every series in a payload is aligned to this one axis by
// construction, so no consumer can plot two panels against different clocks.
type WindowJSON struct {
	Start      int64   `json:"start"`
	End        int64   `json:"end"`
	Period     int32   `json:"period"`
	Range      string  `json:"range"`
	Timestamps []int64 `json:"timestamps"`
}

// JSON renders the window for the wire.
func (w Window) JSON() WindowJSON {
	return WindowJSON{
		Start:      w.Start.Unix(),
		End:        w.End.Unix(),
		Period:     w.Period.Seconds(),
		Range:      w.Range.String(),
		Timestamps: w.Timestamps(),
	}
}

// Payload is what every data endpoint returns. Single-panel endpoints return
// one panel; page endpoints return several — all under one Window, which is the
// mechanism that stops two panels in a response from describing different spans.
type Payload struct {
	Window   WindowJSON `json:"window"`
	Panels   []*Panel   `json:"panels"`
	Warnings []string   `json:"warnings,omitempty"`
}

// NewPayload starts a payload for w.
func NewPayload(w Window) *Payload {
	return &Payload{Window: w.JSON(), Panels: []*Panel{}}
}

// Add appends a panel.
func (p *Payload) Add(panels ...*Panel) *Payload {
	p.Panels = append(p.Panels, panels...)
	return p
}

// Warn appends a payload-level caveat.
func (p *Payload) Warn(format string, args ...any) {
	p.Warnings = append(p.Warnings, fmt.Sprintf(format, args...))
}

// Validate checks the invariants the frontend relies on. Handlers call it
// before writing a response, and the API tests call it on every fixture, so a
// series that has drifted out of alignment with the time axis fails loudly here
// instead of silently plotting shifted data.
func (p *Payload) Validate() error {
	n := len(p.Window.Timestamps)
	if n == 0 {
		return fmt.Errorf("payload has no timestamps")
	}
	for _, panel := range p.Panels {
		for _, s := range panel.Series {
			if len(s.Values) != n {
				return fmt.Errorf("panel %q series %q has %d values, want %d",
					panel.ID, s.Label, len(s.Values), n)
			}
		}
		if t := panel.Table; t != nil {
			if t.Total < int64(len(t.Rows)) {
				return fmt.Errorf("panel %q table reports total %d below its %d rows",
					panel.ID, t.Total, len(t.Rows))
			}
			if want := t.Total > int64(len(t.Rows)); t.Truncated != want {
				return fmt.Errorf("panel %q table truncated=%v disagrees with total %d over %d rows",
					panel.ID, t.Truncated, t.Total, len(t.Rows))
			}
		}
	}
	return nil
}

File: server/src/domain/series.ts
// 응답의 모양. internal/domain/series.go 의 이식이다.
//
// 화면에 뜨는 모든 숫자가 여기를 지난다. 프론트는 포맷팅만 하고 다시 집계하지
// 않는다 — 이전 구현은 표시값을 뷰에서 재집계했고, 그래서 "요청 수" 가 두 곳에서
// 서로 다른 모집단을 셌다.
//
// 클래스마다 toJSON 을 두는 이유는 Go 의 omitempty 때문이다. 빈 값을 그냥 실어
// 보내면 `"color":""` 같은 키가 늘어 두 엔진의 응답이 바이트로 갈린다.

import type {
	Bars,
	Column,
	Intent,
	Panel as PanelJSON,
	Payload as PayloadJSON,
	Point,
	Row,
	Series as SeriesJSON,
	Stat,
	Table as TableJSON,
	Unit,
	WindowJSON
} from '../contract.ts';
import { buckets, periodSeconds, timestamps, compactDuration, type Window } from './window.ts';

export type { Bars, Column, Intent, Point, Row, Stat, Unit };

// 선 무늬. 실선이 기본값인 것은 의도다 — 자기 무늬에 대해 아무 말도 하지 않는
// 시리즈는 이것이 생기기 전에 모든 패널이 그리던 선을 그대로 얻는다.
export const dashSolid = '';
export const dashDashed = 'dashed';
export const dashDotted = 'dotted';

const variantDashes = [dashSolid, dashDashed, dashDotted];

/**
 * variantDash 는 패널 안에서 i 번째 메트릭에 제 나름의 선 무늬를 준다. 같은
 * 주체의 두 메트릭이 한 색을 유지하면서도 구분되게 하려는 것이다. 모자라면
 * 돌려 쓴다 — 한 차트에 메트릭이 셋을 넘는 패널은 무늬로 고칠 수 없는 이유로
 * 이미 읽을 수 없다.
 */
export function variantDash(i: number): string {
	const n = i < 0 ? 0 : i;
	return variantDashes[n % variantDashes.length] as string;
}

/**
 * Series 는 차트의 선 하나다. 감싸는 payload 의 timestamps 와 인덱스가 하나씩
 * 맞물린다.
 *
 * 값이 null 인 것은 진짜 구멍이고 uPlot 까지 null 로 간다 — 선이 끊겨 그려진다.
 * 구멍을 0 으로 메우면 "CloudWatch 에 아무것도 없었다" 가 "값이 0 이었다" 로
 * 바뀌고, 지연 차트에서 그것은 트래픽 붕괴로 읽힌다.
 */
export class Series {
	label: string;
	unit: Unit;
	color: string;
	dash: string;
	values: Point[];

	constructor(label: string, unit: Unit, color: string, n: number) {
		this.label = label;
		this.unit = unit;
		this.color = color;
		this.dash = dashSolid;
		this.values = new Array<Point>(n).fill(null);
	}

	/** 범위 밖 인덱스는 무시한다. 엉뚱한 CloudWatch 타임스탬프가 핸들러를 죽이지 못하게. */
	set(i: number, v: number): void {
		if (i < 0 || i >= this.values.length) return;
		this.values[i] = v;
	}

	/** 칸에 누적한다. 구멍은 0 으로 본다. 여러 원본 행이 한 칸에 떨어지는 개수에 쓴다. */
	add(i: number, v: number): void {
		if (i < 0 || i >= this.values.length) return;
		const cur = this.values[i];
		this.values[i] = cur === null || cur === undefined ? v : cur + v;
	}

	/** 구멍이 아닌 표본의 수. */
	defined(): number {
		return this.values.reduce<number>((n, v) => (v === null ? n : n + 1), 0);
	}

	// max·min·sum·avg·last 는 시리즈를 숫자 하나로 줄인다. 전부 구멍이면 null 이다.
	//
	// 대시보드의 모든 대표 숫자가 이 중 하나를 지난다. 그게 요점이다 — 오버뷰
	// 타일과 상세 차트가 같은 시리즈의 같은 축약을 읽으므로, 이전 구현의 손으로
	// 쓴 클라이언트 축약들처럼 어긋날 수가 없다.
	max(): Point {
		return this.reduce(Math.max);
	}
	min(): Point {
		return this.reduce(Math.min);
	}
	sum(): Point {
		return this.reduce((a, b) => a + b);
	}

	avg(): Point {
		let sum = 0;
		let n = 0;
		for (const v of this.values) {
			if (v === null) continue;
			sum += v;
			n++;
		}
		return n === 0 ? null : sum / n;
	}

	/**
	 * last 는 가장 최근의 정의된 표본이다. newWindow 가 끝을 주기 경계로
	 * 내리므로 마지막 칸은 완성된 칸이다 — 이전 구현이 제 반쯤 찬 꼬리를
	 * 피하려고 쓰던 "끝에서 두 번째 칸" 같은 편법이 필요 없다.
	 */
	last(): Point {
		for (let i = this.values.length - 1; i >= 0; i--) {
			const v = this.values[i];
			if (v !== null && v !== undefined) return v;
		}
		return null;
	}

	private reduce(f: (a: number, b: number) => number): Point {
		let acc: number | null = null;
		for (const v of this.values) {
			if (v === null || v === undefined) continue;
			acc = acc === null ? v : f(acc, v);
		}
		return acc;
	}

	// 키 순서는 Go 의 구조체 선언 순서를 따른다: label, unit, color, dash, values.
	toJSON(): SeriesJSON {
		return {
			label: this.label,
			unit: this.unit,
			...(this.color !== '' ? { color: this.color } : {}),
			...(this.dash !== '' ? { dash: this.dash } : {}),
			values: this.values
		};
	}
}

/** newSeries 는 구멍 n 개짜리 시리즈를 만든다. 인덱스로 채워 넣으면 된다. */
export function newSeries(label: string, unit: Unit, color: string, n: number): Series {
	return new Series(label, unit, color, n);
}

/**
 * newTable 은 행과, 그 행과 무관하게 센 총계를 짝지어 준다.
 *
 * total 은 rows 에서 유도하지 않는다. 이전 구현은 SQL 이 300개로 자른 배열의
 * length 를 총계로 표시했고, 그래서 대표 숫자가 300에서 조용히 세기를 멈추고
 * 옆의 집계와 어긋났다. 호출자에게 total 을 따로 내라고 요구하는 것이 그 모양의
 * 버그를 다시 쓰지 못하게 막는다.
 */
export function newTable(
	columns: Column[],
	rows: Row[] | null,
	total: number,
	limit: number
): TableJSON {
	const list = rows ?? [];
	return { columns, rows: list, total, truncated: total > list.length, limit };
}

/** Panel 은 대시보드의 카드 하나다: 차트, 대표 숫자, 그리고 선택적으로 표. */
export class Panel {
	id: string;
	title: string;
	series: Series[] = [];
	stats: Stat[] = [];
	table: TableJSON | null = null;
	bars: Bars | null = null;
	warnings: string[] = [];

	constructor(id: string, title: string) {
		this.id = id;
		this.title = title;
	}

	/** UI 가 삼키지 말고 드러내야 할 단서를 붙인다. */
	warn(message: string): void {
		this.warnings.push(message);
	}

	toJSON(): PanelJSON {
		const out: PanelJSON = { id: this.id, title: this.title };
		if (this.series.length > 0) out.series = this.series.map((s) => s.toJSON());
		if (this.stats.length > 0) out.stats = this.stats;
		if (this.table !== null) out.table = this.table;
		if (this.bars !== null) out.bars = this.bars;
		if (this.warnings.length > 0) out.warnings = this.warnings;
		return out;
	}
}

/** windowJSON 은 창을 와이어 형태로 만든다. */
export function windowJSON(w: Window): WindowJSON {
	return {
		start: Math.trunc(w.start / 1000),
		end: Math.trunc(w.end / 1000),
		period: periodSeconds(w.period),
		range: compactDuration(w.range),
		timestamps: timestamps(w)
	};
}

/**
 * Payload 는 모든 데이터 엔드포인트가 내는 것이다. 단일 패널 엔드포인트는 패널
 * 하나를, 페이지 엔드포인트는 여럿을 내되 — 전부 하나의 창 아래다. 한 응답의 두
 * 패널이 서로 다른 구간을 설명하는 일을 막는 장치가 그것이다.
 */
export class Payload {
	window: WindowJSON;
	panels: Panel[] = [];
	warnings: string[] = [];

	constructor(w: Window) {
		this.window = windowJSON(w);
	}

	add(...panels: Panel[]): this {
		this.panels.push(...panels);
		return this;
	}

	warn(message: string): void {
		this.warnings.push(message);
	}

	toJSON(): PayloadJSON {
		const out: PayloadJSON = { window: this.window, panels: this.panels.map((p) => p.toJSON()) };
		if (this.warnings.length > 0) out.warnings = this.warnings;
		return out;
	}
}

/**
 * validatePayload 는 프론트가 기대는 불변식을 확인한다. 핸들러가 응답을 쓰기
 * 전에 부르므로, 시간축과 어긋난 시리즈는 조용히 밀린 데이터를 그리는 대신
 * 여기서 큰 소리로 실패한다.
 */
export function validatePayload(p: Payload): void {
	const n = p.window.timestamps.length;
	if (n === 0) throw new Error('payload has no timestamps');

	for (const panel of p.panels) {
		for (const s of panel.series) {
			if (s.values.length !== n) {
				throw new Error(
					`panel "${panel.id}" series "${s.label}" has ${s.values.length} values, want ${n}`
				);
			}
		}
		const t = panel.table;
		if (t !== null) {
			if (t.total < t.rows.length) {
				throw new Error(
					`panel "${panel.id}" table reports total ${t.total} below its ${t.rows.length} rows`
				);
			}
			const want = t.total > t.rows.length;
			if (t.truncated !== want) {
				throw new Error(
					`panel "${panel.id}" table truncated=${t.truncated} disagrees with total ${t.total} over ${t.rows.length} rows`
				);
			}
		}
	}
}

/** newPayload 는 w 를 위한 payload 를 시작한다. */
export function newPayload(w: Window): Payload {
	return new Payload(w);
}

/** 창의 칸 수. 시리즈를 만들 때 길이로 쓴다. */
export function bucketsOf(w: Window): number {
	return buckets(w);
}
